### PR TITLE
Allow to specify a read size in readLp

### PR DESCRIPTION
**Pull request by sinkingsugar**
Also fix secio fragments + test

### CODE REVIEW

File: libp2p/connection.nim [hunk 1]
```diff
@@ -113,7 +113,7 @@ method close*(s: Connection) {.async, gcsafe.} =
     s.isClosed = true
   trace "connection closed", closed = s.closed
 
-proc readLp*(s: Connection): Future[seq[byte]] {.async, gcsafe.} =
+proc readLp*(s: Connection; maxSize: int = DefaultReadSize): Future[seq[byte]] {.async, gcsafe.} =
```
Comment (dryajov) [inline libp2p/connection.nim]: We don't need this, `readLp` means `read length prefixed`, there are other methods such as `read` or `readExactly` that allow you to specify the size. Modifying `readLp` to take a length defeats it's purpose.

Comment (sinkingsugar) [reply libp2p/connection.nim]: Hmm true
Alright, I will just keep the secio fixes and merge those, I will rewrite the test using `read`

**Comment by sinkingsugar** [comment]
I reverted the readLp changes, only a secio fix left so I'm merging this

**Comment by arnetheduck** [comment]
https://github.com/status-im/nim-libp2p/pull/118/files#diff-43d36f9fd0a52d64fd707e3802cae10eR91 - specifying size in `readLp` is better - the API is unsafe without it and we have bugs because of it - if we don't want to refactor the tests, the above is one way to solve it

**Comment by dryajov** [comment]
> https://github.com/status-im/nim-libp2p/pull/118/files#diff-43d36f9fd0a52d64fd707e3802cae10eR91 - specifying size in `readLp` is better - the API is unsafe without it and we have bugs because of it - if we don't want to refactor the tests, the above is one way to solve it

How so, `readLp` is a helper method and it's purpose is to allow reading length prefixed streams, if you don't want to use the helper and you know the size, you can always use `readExactly`.

**Comment by arnetheduck** [comment]
Look at the linked code - `readLp` should take a max size - each framed protocol should have one or the framed protocol is not practically implementable.

**Comment by dryajov** [comment]
I would say then, that this should be per connection object, not a parameter to `readLp`.

**Comment by arnetheduck** [comment]
different protocols have different max frame sizes (mplex: 1mb, secio: 8mb) - sounds complicated to keep state for that.

**Comment by arnetheduck** [comment]
there are already bugs related to this - I'd say the API is bad as is - putting it in state, far from the "problem", just hides it to the implementer of a protocol for no good reason - they'll use a default until they discover a bug just like these bugs are there in our protocols.

**Comment by dryajov** [comment]
It doesn't, this are connection wide settings, why would you have the same connection using different frame sizes, it doesn't make sense and it will lead to even more errors. The frame size should be per connection, otherwise there are going to be constant issues with what the current frame size is for each `readLp` invocation.

**Comment by sinkingsugar** [comment]
>The frame size should be per connection, otherwise there are going to be constant issues with what the current frame size is for each `readLp` invocation.

That's true 👍

**Comment by arnetheduck** [comment]
I guess I think of `readLp` as not really being needed because reading a length-prefixed value is not something inherently tied to a `Connection` - as such, I don't think it should live on the `Connection` interface at all, and neither should any state be there that's needed to correctly and safely implement a frame reading helper.

For example, not all connection types use `readLp` - secio/noise have fixed-length prefixes - should there be a `readXxx` for every type of framing?

I see `readLp` and its friends as stand-alone helpers that operate on a stream and depending on the capabilities of that  stream, might be implemented in different ways (ie if it's a buffered stream, there's one way to implement it and if it's an unbuffered stream, another..).

I also see that whoever calls something like `readLp` should be aware of the fact that it's a DoS vector - a way of highlighting the issue is including the parameter at the call site - that gives the caller a natural signal about how calling `readLp` might fail - ie because someone tried to send too big of a buffer.

Callers will typically not use "different" parameters - they'll use a constant that's defined by the protocol. I guess the constant could be passed to `Connection` on `init` - we can achieve the same "educational" effect by making it a required parameter - but again, that's a bit odd because not all connections need it (a TCP connection that's "just" reading bytes for example).

**Comment by dryajov** [comment]
The thing is that the `Connection` type in libp2p always operates on length prefixed types, otherwise there isn't any need for a `Connection`, you can get away with just any base  `LPStream` types. 

Maybe `SecureConnection` should instead inherit from `LPStream`, I'm impartial there. 

If you look at the implementation of `Connection` you'll see that the only thing it adds are the length prefixed methods `readLp` and `writeLp`, beyond that it does nothing more. I'm fine with moving this methods out to some helper type and just have it operate on `LPStream`, but IMO this will make things harder to follow, at least for me and my mental model.

**Comment by arnetheduck** [comment]
I'm not sure - I interpreted Connection as something keeping track of multiple transports for a single peer - reading lp messages seems outside of its scope - I see reading a length-prefixed message as a general stream utility - then you can have different flavours of length-prefix: varint, bigendian, littleendian etc

**Comment by dryajov** [comment]
`Connection` was initially a utility type to augment `LPStream` with the length prefixed methods, that's it. I agree that this type of methods should probably be part of a higher level streaming API. 

I would say we're at a crossroads in terms of what needs to happen with our streams in general. libp2p streams where modeled after chronos streams, which in hindsight was probably the wrong thing to do. So we're now at a point where we need to rethink this interfaces. I would suggest we work with two abstraction layers (@arnetheduck, I think you've already hinted at this in your `wip` branches). The two interfaces can be something like this:

- A low level interface for reading and writing. Reading can be done with a predicate as you've suggested, that way we use a pull based approach. This low level interface should be super simple but should support building a higher level streaming primitive
- A higher level streaming interface that implements things like, byte ordered reading/writing, length prefixes, etc, that uses the low level interface as a `source` and a `sync` abstraction.

**Comment by arnetheduck** [comment]
If we think about the stream as a stream of objects, there will indeed be many transformation steps in between and some of them could be construed as "levels":

* We start with a stream of bytes - ie the "objects" are individual bytes - this is the most simple stream, and at this level, for practical efficiency, we need to be able to "batch" the bytes into arrays of bytes - so here we're dealing with batches of bytes - `seq` or `openArray`
* Some "filters" turn the stream of bytes into frames - the objects are now "frames" - at this level, the batching is not as urgent.
* The filters might perform operations on those "frames" and return a stream of "bytes" again as they de-onion the payload
* Some filters combine objects of different types: a header, a length and a value - so within the high-level "object", there are lower-level "objects" of different types - it's good if the API can accomodate this and offer support for "combining" predicates or filters, but this I would consider "step 2" - a fully general solution here might be overkill - ie I considered adding functional combinators for the predicates in the wip branch, but doing so would require more "core" functional support and the benefits would not show until we had more advanced filtering cases - I'd keep this in view but not bother too much with it right now.

Crucially though, the implementation of these "filters" would ideally be independent of the type of stream it's operating on and merely specify a "minimal" interface that it needs: `readOnce` ("give me more bytes, but I'll remember all you last gave me") - `readMessage` ("give me more bytes, but you remember what you last gave me until I have all I need") - etc.

The `wip` branch shows what can be done with a "minimal" buffered stream - we could equally easy convert the chronos "native" streams into a `StreamBuffer` with a single `readOnce` predicate and reimplement `readLine` etc using the same code as here.

On top of that, some libp2p types might offer "shortcuts" to certain functions - for example, `readLp` on `Connection` could remain and call a "generic" implementation but perhaps fill in some blanks with the aim to provide a "simplified" API to users so they don't need to bother with the details unless they want / have to - thus there's a difference when talking about the implementation strategy and the end-user API.

**Comment by dryajov** [comment]
This might be interesting as a model for filters and processors - https://pull-stream.github.io/. This also uses a pull based approach (hence the name).

**Pull request merged**